### PR TITLE
Work around build issues with CUDA, MPI, and CMake 3.20.

### PR DESCRIPTION
**Pull request by arthurp**
This is a hack. I'm not pleased, but I cannot currently figure out a way to prevent the generation phase from aborting on 3.20.

This PR is kind of a cry for help. I could any advice would be appreciated. I also asked on the cmake discuss: https://discourse.cmake.org/t/generating-linker-command-line-arguments-to-a-file/3117.

### CODE REVIEW

File: python/katana_setup.py [hunk 1]
```diff
@@ -143,11 +143,44 @@ def load_lang_config(lang):
         return dict(compiler=[], linker=[], extra_compile_args=[], extra_link_args=[], include_dirs=[])
     parser = configparser.ConfigParser()
     parser.read(filename, encoding="UTF-8")
+    linker_wrapper_flag = split_cmake_list(parser.get("cmake", "LINKER_WRAPPER_FLAG", fallback="-Wl,"))
+    linker_wrapper_flag_sep = parser.get("cmake", "LINKER_WRAPPER_FLAG_SEP", fallback=",")
+
```
Comment (ddn0) [inline python/katana_setup.py:171]: probably a short description of the *ahem* workaround and maybe an example here to help future readers orient themselves, e.g.,
"""
...KatanaPythonSetupSubdirectory.cmake generates a file that contains build and link flags that we read back here.... They typically look like: ....

TODO(amp): Starting with CMake 3.20, these flags contain `LINKER:` and `SHELL:` generator expressions which we don't know how to parse generally.
"""

Comment (ddn0) [reply python/katana_setup.py:171]: BTW, there was an update on the discourse issue you raised: https://discourse.cmake.org/t/generating-linker-command-line-arguments-to-a-file/3117/2 though I don't think the the answer is super satisfying.

Comment (arthurp) [reply python/katana_setup.py:171]: Yes. I saw it. It just made me hate CMake. The generator layer is so opaque and almost impossible to get useful information out of except by running the build.

File: cmake/Modules/KatanaPythonSetupSubdirectory.cmake [hunk 2]
```diff
@@ -99,27 +111,33 @@ function(_generate_build_configuration_ini)
...
-    list(APPEND COMPILE_DEFINITIONS "$<TARGET_PROPERTY:${TARGET_NAME},INTERFACE_COMPILE_DEFINITIONS>")
-    list(APPEND LINK_OPTIONS "$<TARGET_PROPERTY:${TARGET_NAME},INTERFACE_LINK_OPTIONS>")
-    list(APPEND LINK_OPTIONS "-Wl,-rpath=$<TARGET_LINKER_FILE_DIR:${TARGET_NAME}>")
-    list(APPEND LINK_OPTIONS "-Wl,-rpath=$<JOIN:$<TARGET_PROPERTY:${TARGET_NAME},BUILD_RPATH>,;-Wl,-rpath=>")
-    list(APPEND LINK_OPTIONS "-Wl,-rpath=$<JOIN:$<TARGET_PROPERTY:${TARGET_NAME},INSTALL_RPATH>,;-Wl,-rpath=>")
-    list(APPEND LINK_OPTIONS "$<TARGET_LINKER_FILE:${TARGET_NAME}>")
-    list(APPEND COMPILE_OPTIONS "$<TARGET_PROPERTY:${TARGET_NAME},INTERFACE_COMPILE_OPTIONS>")
+    list(APPEND INCLUDE_DIRECTORIES "$<TARGET_GENEX_EVAL:${TARGET_NAME},$<TARGET_PROPERTY:${TARGET_NAME},INTERFACE_INCLUDE_DIRECTORIES>>")
+    list(APPEND COMPILE_DEFINITIONS "$<TARGET_GENEX_EVAL:${TARGET_NAME},$<TARGET_PROPERTY:${TARGET_NAME},INTERFACE_COMPILE_DEFINITIONS>>")
+    # TODO(amp): Including INTERFACE_LINK_OPTIONS causes an unavoidable cmake generation failure in cmake 3.20+ due to the user of $<HOST_LINK:...>.
```
Comment (ddn0) [inline cmake/Modules/KatanaPythonSetupSubdirectory.cmake]:
```
"due to the user of" =>
"due to the use of"
```

File: cmake/Modules/KatanaPythonSetupSubdirectory.cmake [hunk 3]
```diff
@@ -99,27 +111,33 @@ function(_generate_build_configuration_ini)
...
+    #  INTERFACE_LINK_OPTIONS doesn't seem to be require for the dependencies we are using right now, so just skip it for the moment.
```
Comment (ddn0) [inline cmake/Modules/KatanaPythonSetupSubdirectory.cmake]:
```
"doesn't seem to be require" =>
"doesn't seem to be required"
```

File: python/katana_setup.py [hunk 4]
```diff
@@ -138,17 +138,72 @@ def test_cython_module(name, cython_code, python_code="", extension_options=None
...
+    Load the compilation configuration provided by CMake.
+
+    KatanaPythonSetupSubdirectory.cmake generates a JSON file that contains build and link flags that we read.
+    They typically look like:
+
+    .. code-block:: json
+
+        {
+        "COMPILER":"ccache;/home/amp/LocalInstalls/conda/miniconda3/envs/katana-dev/bin/clang++",
+        "INCLUDE_DIRECTORIES":"/home/amp/shared/Katana/katana-enterprise/libquery/include;/home/amp/shared/Katana/katana-enterprise/external/katana/libgalois/include;/home/amp/shared/Katana/katana-enterprise/external/katana/libtsuba/include;/home/amp/shared/Katana/katana-enterprise/external/katana/libsupport/include;/home/amp/shared/Katana/katana-enterprise/build-tests/external/katana/libsupport/include;/home/amp/LocalInstalls/conda/miniconda3/envs/katana-dev/include;/home/amp/LocalInstalls/conda/miniconda3/envs/katana-dev/lib/backward;/home/amp/shared/Katana/katana-enterprise/libgluon/include;/home/amp/shared/Katana/katana-enterprise/libdist/include;/home/amp/shared/Katana/katana-enterprise/libtsuba-fs/include;/home/amp/shared/Katana/katana-enterprise/external/katana/libtsuba/src",
```
Comment (ddn0) [inline python/katana_setup.py]: suggestion: feel free to elide bits and just keep the `LINKER` and `SHELL` portions:

```
.. code-block:: json
    {
     "COMPILER": ....
     "INCLUDE_DIRECTORIES": .....
     "LINK_OPTIONS": <as before>
     "LINKER_WRAPPER_FLAG": <as before>
     "LINKER_WRAPPER_FLAG_SEP": <as before>
    }
```

**Review by ddn0 — APPROVED** [review]

**Pull request merged**